Add nullable option to attributes (#81)
talyssonoc/structure: docs/schema-concept/nullable-attributes.md

# Nullable attributes

> You can change the way an attribute is treated when the value `null`is assigned to it by using the `nullable` option with the value `true`, this would affect the way the attribute is coerced, validated and serialized.

Reviewer: Missing one space here
```diff
- value `null`is
+ value `null` is 
```
:)

If you do not set the `nullable` option for an attribute it will default to `false` and make your attribute __not nullable__.

> 

Reviewer 2: What do you think about adding a text here:

```suggestion
Here's the behavior without the `nullable` option:
```

```javascript
/*
 * User.js
*/
const User = attributes({
  name: {
    type: String,
    empty: true
  },
  age: Number,
  active: Boolean,
  createdAt: Date
})(class User { });

const user = new User({
  name: null,
  age: null,
  active: null,
  createdAt: null
});

user.attributes; // { name: '', age: 0, active: false, createdAt: 1970-01-01T00:00:00.000Z }
user.validate() // { valid: true }
```

Notice that the __default__ value for __String__ is empty, what it means that you need to accept empty values to turn your schema valid.

### Accepting nullables (optional attributes)

When you choose to accept __nullables__ in attributes that you actually not require our coercion functions will just not define a __default__ value for those attributes and will assign `null` to them.

```javascript
/*
 * User.js
*/
const User = attributes({
  name: {
    type: String,
    nullable: true
  }
})(class User { });

const user = new User({
  name: null
});

user.attributes; // { name: null }
user.validate() // { valid: true }
```

### Accepting nullables (required attributes)

> We understand that when an attribute is required it should to be present and actually has a value assigned to it, even an `undefined`, `null` or whatever value what it means that coercion function will not assign a __default__ value even if __nullable__ attribute is __false__.

Reviewer 2: ```suggestion
We consider that when an attribute is __required___ there should be some value assigned to it even if it's `undefined`, `null` or any other value. It means that coercion will never assign a __default__ value to __required__ attributes even if __nullable__ option is __false__.
```

```javascript
/*
 * User.js
*/
const User = attributes({
  name: {
    type: String,
    required: true,

>     nullable: false

Reviewer 2: I think we can add a comment here like `// non-nullable required attribute`

  }
})(class User { });

const user = new User({
  name: null
});

user.attributes; // { name: null }
user.validate() // { valid: false }
```

> But notice that you can choose to allow __null__ values on __required__ attributes which will cause the validation to return __true__.

Reviewer 2: ```suggestion
But notice that you can choose to allow __null__ to be a valid value for __required__ attributes by making it __nullable__, which will cause the validation to return __true__.
```

```javascript
/*
 * User.js
*/
const User = attributes({
  name: {
    type: String,
    required: true,
    nullable: true
  }
})(class User { });

const user = new User({
  name: null
});

user.attributes; // { name: null }
user.validate() // { valid: true }
```

**Important:**

> - Notice that usually an __undefined__ or __null__ value is not returned when you serialize your schema but when you accept a __nullable__ attribute this attribute is going also to be returned in your serialized schema, see [Serialization](serialization.md)

Reviewer 2: ```suggestion
- Notice that usually an attribute with the value __undefined__ or __null__ is not included when you serialize your structure, but when it is __nullable__ this attribute is going also to be returned in your serialized schema.
```

> - Notice that usually an attribute with the value __undefined__ or __null__ is not included when you serialize your structure, but when it is __nullable__ this attribute is going to be returned in your serialized schema.

Reviewer 2: What do you think?

```suggestion
- Notice that usually an attribute with the value __null__ is not included when you serialize your structure, but when it is __nullable__ this attribute is going to be returned in your serialized structure even if its value is __null__.
```
